src/common/ServerNotifsExpiringBanner.js

```diff
@@ -0,0 +1,76 @@
+/* @flow strict-local */
+
+import React from 'react';
+import type { Node } from 'react';
+import subWeeks from 'date-fns/subWeeks';
+
+import ZulipBanner from './ZulipBanner';
+import { useSelector, useDispatch, useGlobalSelector } from '../react-redux';
+import { getAccount, getSilenceServerPushSetupWarnings } from '../account/accountsSelectors';
+import { dismissServerNotifsExpiringBanner } from '../account/accountActions';
+import {
+  kPushNotificationsEnabledEndDoc,
+  pushNotificationsEnabledEndTimestampWarning,
+} from '../settings/NotifTroubleshootingScreen';
+import { useDateRefreshedAtInterval } from '../reactUtils';
+import { openLinkWithUserPreference } from '../utils/openLink';
+import { getGlobalSettings } from '../directSelectors';
+
+type Props = $ReadOnly<{||}>;
+
+/**
+ * A "nag banner" saying the server will soon disable notifications, if so.
+ *
+ * Offers a dismiss button. If this notice is dismissed, it sleeps for a
+ * week, then reappears if the warning still applies.
+ */
+export default function ServerNotifsExpiringBanner(props: Props): Node {
+  const dispatch = useDispatch();
+
+  const globalSettings = useGlobalSelector(getGlobalSettings);
+
+  const lastDismissedServerNotifsExpiringBanner = useSelector(
+    state => getAccount(state).lastDismissedServerNotifsExpiringBanner,
+  );
+
+  const perAccountState = useSelector(state => state);
+  const dateNow = useDateRefreshedAtInterval(60_000);
+
+  const expiryWarning = pushNotificationsEnabledEndTimestampWarning(perAccountState, dateNow);
+
+  const silenceServerPushSetupWarnings = useSelector(getSilenceServerPushSetupWarnings);
+
+  let visible = false;
+  let text = '';
+  if (expiryWarning == null) {
+    // don't show
+  } else if (silenceServerPushSetupWarnings) {
+    // don't show
+  } else if (
+    lastDismissedServerNotifsExpiringBanner !== null
+    && lastDismissedServerNotifsExpiringBanner >= subWeeks(dateNow, 1)
+  ) {
```

Comment:
So I wonder if we should make the time window that a dismissal lasts substantially longer for non-administrators than it is for administrators? Like maybe 2 months for normal users? We probably do want it to be finite, since in theory there could be separate situations 6 months apart that cause notifications to be potentially expiring. But I think it's mostly administrators and owners who would benefit from an additional reminder of the impending change before it actually happens, and I can see it being frustrating for people who might be in the middle of their procurement process for all their employees to get the notice a second time.